api/src/handlers/websocket/used-ports.go

```diff
@@ -0,0 +1,112 @@
+package websocket
+
+import (
+	"porter/api/src/handlers/websocket/message"
+	"fmt"
+	"porter/api/src/lib/github"
+	"porter/api/src/lib/redis"
+	"bytes"
+	"porter/api/src/lib/dockerfile"
+	"strings"
+	"time"
+)
+
+const redisCacheKey = "PORTER:REPOSITORIES"
+
+func exposedPortLookup(repository github.RepositoryResponse,  msgCh chan message.Message) (github.Repository)  {
+	ports := make([]string, 0)
+
+	message.SendMessage(fmt.Sprintf( "[%s] - fetching Dockerfile", repository.Name),"green", msgCh, false)
+	body, err := github.FetchFileContent(repository.Name, "Dockerfile")
+	if err != nil || body == "" {
+		message.SendMessage(fmt.Sprintf( "[%s] - couldn't get Dockerfile", repository.Name),"red", msgCh, false)
+	} else {
+		buf := bytes.NewBuffer([]byte(body))
+		message.SendMessage(fmt.Sprintf( "[%s] - parsing Dockerfile into an AST", repository.Name),"green", msgCh, false)
+		port, _ := dockerfile.ExtractExposedPort(buf)
+
+		if port != "" {
+			message.SendMessage(fmt.Sprintf( "[%s] - found the port %d in the Dockerfile", port, repository.Name),"green", msgCh, false)
+			ports = append(ports, port)
+		}
+
+		message.SendMessage(fmt.Sprintf( "[%s] - did not found any port in the Dockerfile", repository.Name),"red", msgCh, false)
+	}
+
+	if len(ports) == 0 {
+		message.SendMessage(fmt.Sprintf( "[%s] - searching ports in k8 config files", repository.Name),"green", msgCh, false)
+		k8Ports := github.FetchK8Port(repository)
+
+		portsLength := len(k8Ports)
+
+		if portsLength == 0 {
+			message.SendMessage(fmt.Sprintf( "[%s] - did not find any port in k8 config files", repository.Name),"red", msgCh, false)
+		} else {
+			message.SendMessage(fmt.Sprintf( "[%s] - found ports in k8 config files %s", repository.Name, strings.Join(k8Ports, ",")),"green", msgCh, false)
+		}
+
+		for i:= 0; i<portsLength; i++  {
+			ports = append(ports,k8Ports[i])
+		}
+	}
+
+	return github.Repository{
+		ExposedPorts: ports,
+		Name: repository.Name,
+		Url: repository.Url,
+	}
+}
+
+func usedPortsHandler(ch chan github.Repository, msgCh chan message.Message) {
+	message.SendMessage("fetching repositories in parallel","green" , msgCh, false)
+	repositories := github.FetchAllRepositories()
+
+	if len(repositories) == 0{
+		message.SendMessage("Aborting session, could not find repositories", "red", msgCh, true)
+		close(ch)
+	}
+
+	message.SendMessage(fmt.Sprintf( "Done fetching repositories, %d total", len(repositories)),"green" , msgCh, false)
+	repositoriesPortCh := make(chan github.Repository, len(repositories))
+
+	message.SendMessage(fmt.Sprintf( "fetching results from cache (redis)"),"green" , msgCh, false)
+
+	redisClient := redis.GetClient()
+	cachedResults := redisClient.HGetAll(redisCacheKey)
+	redisHashUpdated := false
+	for i := 0; i < len(repositories) ; i++ {
+		go func(i int) {
```

Comment:
your code is very nested and cluttered. break to it functions with better naming so the top functions can be read easily.
```go
func do() {
  a()
  if !true {
    b()
  }
}
```